docs: Add machine.CAN low-level API docs.

projectgus opened the pull request:
First step to implementing a cross-port set of machine.CAN drivers, as outlined in #12337. Refer to the linked PR for the whole plan.

**********************************************************************************************
**This is a draft of a new API at this point. There might still be significant changes before it can be accepted.**
**********************************************************************************************

## Outline

The proposed API implements `machine.CAN` as a low-level hardware abstraction layer that provides three things:

1. A priority hardware queue of outgoing CAN messages.
2. A FIFO hardware queue of incoming CAN messages, with optional hardware filters.
3. Bus error & state change reporting and control.

`machine.CAN` is enough to do some *very basic* CAN operations, even more basic than `pyb.CAN`. The idea is to layer `micropython-lib` modules for `can` and `aiocan` on top of `machine.CAN`. These modules will provide extra functions like software message queues, a more "Pythonic" API, etc.

The `machine.CAN` API is also simple enough that it should be possible to implement pure Python drivers for external CAN controllers like MCP2515, MCP2518FD, etc. and use these with `can` and `aiocan`.

## can & aiocan

Currently this higher layer in micropython-lib (*not part of this PR*) is less well defined, but I''d like it to look quite similar to python-can, at least for the synchronous API. Anticipating the code in the higher layer will look something like:

```py
import can
import machine

mcan = machine.CAN(0, bitrate=1_000_000, mode=machine.CAN.NORMAL)
bus = can.Bus(mcan, send_depth=8, recv_depth=16)  # Set explicit software message queue depths

msg = can.Message(0x300, data=bytearray(8), is_extended=False)
counter = 0

def send_message(_timer):
    global counter
    counter += 1
    struct.pack_into("I", msg.data, 0, counter)
    bus.send(msg)

tim = machine.Timer(-1)
tim.init(period=100, callback=send_message)

r_msg = None

while True:
    r_msg = bus.recv(timeout=None, message=r_msg)  # reuse Message object
    if r_msg.identifier < 0x300:
       print(hex(r_msg.identifier, r_msg.data.hex())
```

Or async:

```py
import asyncio
import aiocan
import machine

mcan = machine.CAN(0, bitrate=1_000_000, mode=machine.CAN.NORMAL)
bus = aiocan.Bus(mcan, send_depth=8, recv_depth=16)

bus.filter(aiocan.Filter(message=0x100, mask=0x100, is_extended=False))

async def send():
    msg = aiocan.Message(0x300, data=bytearray(8), is_extended=False)
    counter = 0
    while True:
        await asyncio.sleep_ms(100)
        counter += 1
        struct.pack_into("I", msg.data, 0, counter)
        await bus.send(msg, timeout=None)

async def recv():
    r_msg = None
    while True:
        r_msg = await aio.recv(timeout=None, message=r_msg)
        if r_msg.identifier < 0x300:
           print(r_msg)

async def main():
    await asyncio.gather(send(), recv())

asyncio.run(main())
```

## Limitations

* Dealing with multiple hardware FIFOs is left for the driver to manage. The caller sees incoming CAN messages in FIFO order.
* Filtering may be less capable than what some hardware is able to do.

## Concerns

* Python code may not be a fast enough "message pump" to keep the hardware TX queue full, and the RX queue empty. I think it will be, but remains to be proven.
* CAN controller differences may make a common `machine.CAN` API too difficult to implement, with too many "quirks" of behaviour.
* Not including a `recv()` method (only a callback) in the low-level `machine.CAN` may be too restrictive. However, the goal is for everyone to use the higher level `can` or `aiocan` modules.

## Credit

This API is variously inspired by [MicroPython aioble](https://github.com/micropython/micropython-lib/tree/master/micropython/bluetooth/aioble), [pyb.CAN](https://docs.micropython.org/en/latest/library/pyb.CAN.html), [CANPico](https://raw.githubusercontent.com/kentindell/canhack/master/CANPico/docs/CANPico%20MicroPython%20SDK%20reference%20manual.pdf), and the [Rust bxcan crate](https://docs.rs/bxcan/latest/bxcan/). Misunderstandings and bad takes are 100% my own work, though!

*This work was funded through GitHub Sponsors.*

projectgus committed:
docs: Add machine.CAN API docs.

First stage to implementing a cross-port set of machine.CAN driver.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

projectgus commented:
Regretfully I am putting this up and vanishing on holiday for a few weeks. :beach_umbrella:  I will be back around Jan 2 and will carefully read any comments then. Looking forward to pushing ahead with the CAN API!

robert-hh commented:
> I am putting this up and vanishing on holiday for a few weeks.

This is quite a cliff-hanger ...

kwagyeman commented on docs/library/machine.CAN.rst:
```diff
@@ -0,0 +1,364 @@
+
+   Otherwise, they can be specified:
+
+    - ``sjw`` is the resynchronisation jump width in units of time quanta for nominal bits;
```
I'd remove all this stuff. It's really hard to use except for the STM32. Even then it's hard. The sample_point value gives you enough control as it adjusts tseg1/tseg2 given f_clock. sjw is a valid setting to provide control for as it's independent of the clock freq. But, only certain ports will be able to set it. On the imxrt it can't be set independently of the tseg1/tseg2.

projectgus replied:
Makes sense. I'll look more closely at some of the other controller implementations and see what's the best cross-compatible way to slice it (I see @kdschlosser made a similar recommendation in their comment). Can always have port-specific optional arguments where available.

I feel like >98% of Microython users will only ever set bitrate anyhow.

kwagyeman replied:
Sample point control is important however. This is needed to get CAN devices working on buses when you have "un-ideal" bus setups with long runs coming off a bus.

projectgus replied:
Agree. How about slicing it like this?

* Controller-agnostic arguments for `bitrate` and optional custom sample point.
* Controller-specific arguments to explicitly set more if/when needed, with whatever features that controller exposes.

kdschlosser replied:
This is what I am talking about as far as doing the calculations for a bitrate

https://github.com/kdschlosser/can_timing/blob/main/can_timing/__init__.py

That is an earlier version and I have a more updated version that has more interfaces added to it. But that is the basic idea. standard values fed in and out comes interface specific registers. The user should not need to know how to calculate those registers when all they want to do is connect to a network at a given bitrate.

The other thing that this kind of thing can be used for is connecting to a network without even providing a bitrate. You would use this with the most commonly used bitrates and how it works is the function gets fer in the bitrate returns the registers, the registers get set and then you look for any errors. if there are any errors then there is a pretty good chance that the bitrate being used is not what the network is running at, then move onto the next one. Once valid frames are received and there are no errors then you know you have gotten the correct bitrate.

projectgus replied:
> I'd remove all this stuff. It's really hard to use except for the STM32. 

Looking at these arguments (`sjw`, `tseg1`, `tseg2`) again and comparing the various controller data sheets, I think I'd like to keep them for all controllers. You're right @kwagyeman that most folks won't use them, but they will be useful for fine-tuning parameters in cases like the ones that @kdschlosser is working with.

There seems to be a split in terms of the controller hardware that is configuring using `tseg1 = prop_seg + phase_seg1` (includes: STM32, ESP32, MCP2518FD), and controller hardware that sets `prop_seg` and `phase_seg1` individually (includes: iMXRT, MCP2515).

As per the standard, it's sufficient for a controller to only program the total `tseg1`. So the question becomes which method is more general, and secondarily which method uses less resources.

It's not a clear win either way, but I've kept `tseg1` and `tseg2` mostly to save a parameter. I think drivers for hardware that needs to split this into `prop_seg` and `phase_seg1` should be able to do so pretty easily within the constraints of the controller registers. @kwagyeman does that seem OK to you?

The PR has been updated with some more explanation about this, also.

projectgus replied:
> The only way to keep the API identical across the board for the constructors is going to be using code like this. There is no choice in that. The issue is the interfaces having different registers that need to be set. 

@kdschlosser Just to check, does "code like this" mean code like what's currently proposed in this PR, or code like in the linked `can_timing` project?

kdschlosser replied:
Everything I am talking about is written in C code, nothing is written in Python. The example I supplied was written to run in MicroPython. It's not pretty but it works. it's also a resource hog and that wouldn't be the case if it was written in C code. The amount of space it takes up in terms of program size is going to be really small.

kdschlosser replied:
@projectgus 

code like this means the example in that link I posted. code that is able to take simple understoood easy to explain values and turn them into the timing registers needed for an interface.

kwagyeman replied:
I think it's best to leave the bit rate calculations up to the HAL APIs supplied by each MCU. You only need to control the sample point, bitrate, and sjw to specify everything. As for a tolerance parameter... I think you're goal would always be to hit a specified bitrate and sample point exactly. If you aren't then that would be an error. It's not like people really need the maximum amount of flexibility here.

Anyway, for my purposes I will update the MIMXRT CAN driver to work with whatever is proposed.

projectgus replied:
Thanks @kdschlosser and @kwagyeman for the extra explanation and input. @kdschlosser , I think I understand your reasoning and concerns now. I think we can meet these requirements, although probably with a slightly different design.

> You only need to control the sample point, bitrate, and sjw to specify everything.

I think not quite, though? SJW units are typically time quanta, and this isn't enough information for the caller to know the time quantum length. So would either need to specify SJW as a proportion, or pass additional information here. Let me know if I'm missing a trick here.

---

To try and summarise, I counts five different bit timing methods that have been mentioned somewhere in this PR:

1. Set a nominal `bitrate`, driver chooses the rest. *I'm certain this is all >95% of MicroPython users will ever need, and MP should ensure it is the simplest thing to do.*
2. Set a `bitrate` and a sample point percentage. This allows the driver to calculate everything except for SJW, which will have to be inferred. As @kentindell mentions below, it's easy to get this value wrong.
3. Set `bitrate`, `sjw`, `tseg1` and `tseg2` in units of time quanta. This maps to the CAN ISO standard parameters and (I think) fully defines the bit timings.
4. Set `bitrate`, `sjw`, `prop_seg`, `phase_seg1` and `phase_seg2`. This maps even more closely to the CAN ISO standard, but it's functionally equivalent to 3 (where `tseg1 = prop_seg + phase_seg1`), and the standard even says as much.
5. Provide "first principles" bus physical properties and allowed tolerances, as described above by @kdschlosser. There's an appealing elegance to this, but it's a different approach to what the standard defines which is what most folks working with CAN are more likely to be used to (especially if copying parameters from an existing project, etc.)

At this stage I'm thinking all low-level controller drivers should supports methods 1, 2 & 3. MicroPython can include a generic helper function to calculate method (3) given the inputs from methods 1 & 2 and driver-specific constraints. Method 5 should be possible to implement in Python as a set of functions on top of method 3, and could be bundled in micropython-lib as an addon package.

kwagyeman replied:
Works for me.

kdschlosser replied:
The standard defines using the bus length and the node delay as the way to calculate what the SJW is. Most people will just plug a number in and see how it works because after all there is typically only going to be an SJW of 1 to 4. Trial and error in this case is really simple to do. So no big deal there. Those are things that don't need to be added, it can be done is what I was getting at.

The only things that a user would need to supply for the calculations to do their thing would be the bitrate and the SJW. Those 2 things and they would be off and running. They don't need to know a single thing about the CAN specification or how to calculate tseg or propsegs or the timing registers. All of that is done internally as it should be.  Most people that are working with MCU's have an idea of what the bitrate is and the SJW is really only going to come into play with longer busses or buses that have a nigher node count. For the most part that number will end up being set to a 1. The SJW should be defaulted to 1 making the only thing that the user needs to supply is the bitrate. That's it.

I had also written code that does not iterate over the BRP, TSEG1 and TSEG2 ranges. It uses raw calculations instead. The issue there is it is not going to allow any deviation. by doing the iteration method we can use a close bitrate and it would work because of the ability for can to adjust for things like oscillator float. a lot of times you can go as much as +- 5% of the target bitrate. 

If I asked you to tell me what the timing registers are or what the brp, tseg1 and tseg values are for an ESP32V3 running at 80kbps with a sample point of 87.5% could you do it?. Could you locate a table somewhere in the ESP's manual or in the SDK's manual that would tell you those numbers? I already know the answer which is you are not going to be able to locate those numbers. You are not going to find a calculator that would be able to generate the numbers either.

kdschlosser replied:
If a person has the knowledge of how to get the tseg values and the brp value then they will have the knowledge or at least know how to find how to convert those values into the timing register numbers. This is the reason why I said the only thing that needs to be supplied in terms of an "advanced user" way of doing this is to allow the user to pass the timing registers as a list or tuple to a keyworded parameter "registers" The interface being used is going to know how the length of that passed in tuple or list so it would be able to handle collecting the registers from it without any issues. The list or tuple also allows the common front facing API to be blind as to what is actually being passed. some interfaces only have a single register  some have 2 or 3. The common facing API doesn't need to know what is being passed. That would get handled by the board specific portion.

projectgus replied:
@kdschlosser I've read these latest comments several times and I think I understand them individually, but I honestly don't fully understand what specifically you're asking to change compared to [what's already outlined above](https://github.com/micropython/micropython/pull/13149#discussion_r1448089343) and in the PR.

I did note one suggestion, which was to support initialising with only two parameters: `bitrate` and `sjw`. From your comment I see that "start from sjw=1 and increase until the errors go away" is a reasonable method for finding an SJW value in some use cases, so supporting this seems useful and viable. Thanks, will change.

Regarding exposing hardware-specific register values as an initialisation method, I agree with the comment you made earlier about this being "a support headache". It's reasonable to expose hardware-specific features in initialisation, where they don't have any generalised equivalent, but the driver should abstract this into a common interface as much as possible (similar to the other MicroPython machine drivers). The user who wants to write registers directly always has `machine.mem32` at their disposal.

If I missed or misunderstood an additional API change that you'd like to see, compared to what's already there, then please outline the specific proposed change, and the associated benefit compared to the status quo, as concisely as possible.

kwagyeman commented on docs/library/machine.CAN.rst:
```diff
@@ -0,0 +1,364 @@
+   same meanings and behaviour as that function, however they configure the
+   Bit Rate Switch (BRS) feature of CAN FD.
+
+   When specifying exact timing parameters, the accepted ranges are:
```
Same as comment above. How this works makes sense. On the imxrt I'd have to deinit and then reinit the can though after this was called. But, I could keep the regular settings. As mentioned above, giving tseg1/tseg2 control is something few people will use and it's easier to use the abstracted sample_point api as that's what you are trying to achieve by adjusting the segment times.

projectgus replied:
> giving tseg1/tseg2 control is something few people will use

We can probably roll this into the above comments, with the goal that whatever API is available in `init()` for CAN Classic will be mirrored in `init_fd()` for CAN FD timings.

> On the imxrt I'd have to deinit and then reinit the can though after this was called. But, I could keep the regular settings.

I was just reading over this part in the iMX datasheet. I agree this is a bit fiddly in this case, sorry about that! On balance, it still seems better than trying to jam all the parameters into one place though.

Your comment did also get me thinking through the implications of having a CAN FD controller come online for a short period in an incompatible mode, before it reconfigures to support `BRS`. I think the way around this will be for the high-level `can` Python class to call `init(mode=CAN.Mode.SLEEP, ...)` and then `init_fd(mode=CAN.Mode.NORMAL, ...)` so the "Classic CAN" initialisation doesn't result in a transient period where the controller is active on the bus.

kwagyeman commented on docs/library/machine.CAN.rst:
```diff
@@ -0,0 +1,364 @@
+   - ``tseg1`` can be a value between 1 and 32 inclusive
+   - ``tseg2`` can be a value between 1 and 32 inclusive
+
+.. method:: CAN.detect_bitrate([timeout_ms])
```
Unlikely to be implemented by most ports.

kdschlosser replied:
This can easily be implemented by all ports. I have code I have written that generated the registers used to set bitrates for the majority of the CAN IC's made. it's a mater of iterating over the most common CAN speeds used collecting the registers for the chip that is being used. setting the registers in place. When messages start coming in and there are no errors you have located the correct bitrate. 

You can only do this for the typical bitrates because of how long it would take to run. If the user is supplying a timeout that number can be used to determine what bitrates are going to be tested.

kwagyeman replied:
You don't typically want to cause errors on the CAN bus when turning on... so, you'd need to init the controller very carefully is listen only mode. I guess it's not hard... but, most HAL code from like ST or NXP doesn't allow you to do this easily. You'll have to work from the register set. That said, I guess I had to do the same for the NXP MIMXRT port to implement some of the requested features.

projectgus replied:
I'll remove this. I'm not sure what my motivation for adding it was, I think I saw it in another driver's API and thought "oh, some controllers must support this as a hardware feature". But I they actually don't, and now I can't even find the driver where I saw it(!)

It should be possible to for someone to still add a higher-level function that works as @kdschlosser describes, without it being in the low-level driver API.

projectgus replied:
Removed from the base classes.

kwagyeman commented on docs/library/machine.CAN.rst:
```diff
@@ -0,0 +1,364 @@
+
+.. method:: CAN.send(identifier, data, flags=0, fifo_equal=True)
+
+   - ``identifier`` is an integer CAN identifier value.
```
arg order is reversed from pyb can.

projectgus replied:
Yeah... I find the pyb argument order here very unsettling for some reason, but I guess it's worth keeping it consistent. :)

projectgus replied:
Went back and forth on this but decided to keep it. The remaining arguments are complete incompatible with `pyb.CAN`, so if someone copy-pastes code from there then they're going to have a bad time anyway. At least this way it's more likely to  produce an obvious error quicker, rather than more subtle incorrect behaviour.

kwagyeman reviewed (approved):
API seems to work. I would reconsider adding a recv() method so folks don't have to change their code too much. If the higher level CAN module lands at the same time then I think there's not a problem.

Anyway, after this is approved I can modify the imxrt can implementation to follow this spec pretty easily.

kdschlosser commented:
There should be 2 distinct classes for the CAN. It should be CAN and CANFD. they should not be combined. I would also recommend not using "CAN" or "CANFD" as these are copyrighted by Bosch. I believe the copyright has expired in the US but it has not in Europe and use of those phrases could cause an issue. This is the reason why Espressif uses TWAI.

Regardless of what it is named there needs to be a distinction made between the 2 network types. There should also be a "Message" class. This class represents a CAN frame. It contains the arbitration ID's, the data contained in the packet. along with any other information that is attached to a frame. The information that is going to be available in a message is going to vary based on the flavor and version of CAN that is being used nd if something in a message is not used then it would be set to None. The message class is what gets handed back to a user when collecting the received messages. This is what gets passed to the transmit function. only 2 things should get passed to the transmit function, one of these messages and a timeout if wanting to make a blocking transmit call. If a call is non blocking (user has a registered callback for transmit errors) then timeout is optional and would carry a default value of -1.

Few years back I wrote a program that you can feed in a bitrate and some static limits based on the CAN IC being used and it would return the registers needed to set the bitrate for the CAN IC being used.  This is something that MUST be available. A user really should not have to know things like BRP, TSEG1, TSEG2, SJW, PHSEG1, PHSEG2, TQ, PROPSEG, (CNF1, CNF2 and CNF3), (BTR0 and BTR1), DCANBTR, (BTP and FBTP), (CiCFG1 and CiCFG2), CxBCR, CnBTR, CiCONR, etc.... The user should be able to pass to the constructor the target bitrate, bus length(optional), transceiver timings (optional) and number of nodes (optional) and the software will calculate the required registers needed to make it work. Optionally There can be BRP, TSEG1, TSEG2, SJW and sample point parameters can be used. The software should have the ability to take those optional parameters and turn them into the timing registers for the CAN interface that is being used.

I recommend splitting the CAN implementation into 2 pieces. the first piece is what the user interacts with. it is the classes. The other is the bus. The only exposed portion of the bus is the constructor. The user will not have access to any other portion of the bus. This would allow for MCU's that do not have built in CAN interfaces to be able to connect an external one over I2C or SPI. This would keep the exposed user API identical across the board. The "busses" would need to have specific functions that the CAN classes would call to perform tasks. The API for those functions would be defined as well. The bus would be responsible for carrying out the task that it has been instructed to do and it must do it with the information that has been given. 

The hardest part is the initialization of a bus. There is no standard so the information that is needed by an IC is going to be all over the place based on model and manufacturer of the IC. That is why it is imperative that code like what I wrote is able to calculate the timing registers using a standardized set of inputs. Without having that available you will not be able to have a common API between the different IC's.

projectgus commented:
@kdschlosser Thanks for the in-depth comments here.

>  I would also recommend not using "CAN" or "CANFD" as these are copyrighted by Bosch. I believe the copyright has expired in the US but it has not in Europe and use of those phrases could cause an issue. This is the reason why Espressif uses TWAI.

Perhaps you're thinking of patents and/or trademarks? It's the patent on CAN which has expired. CAN and CAN-FD are still Registered Trademarks of Bosch, and IANAL but [as I understand it](https://www.bosch-semiconductors.com/ip-modules/can-protocol-license/) their licensing concerns are around hardware/gateware implementations and not software drivers for those implementations. For all controllers except for Espressif TWAI, we're writing drivers for CAN controller hardware (presumably with any applicable hardware IP licensing in place), so I think it's reasonable to name the driver as such. (See also: CAN drivers in open source operating systems like Zephyr, Linux, NuttX.) For the particular case of Espressif then I guess we can add some disclaimer that it's not actual CAN controller hardware in that case, and/or the low level part can be `machine.TWAI`.

> A user really should not have to know things like BRP, TSEG1, TSEG2, SJW, PHSEG1, PHSEG2, TQ, PROPSEG, (CNF1, CNF2 and CNF3), (BTR0 and BTR1), DCANBTR, (BTP and FBTP), (CiCFG1 and CiCFG2), CxBCR, CnBTR, CiCONR, etc.... The user should be able to pass to the constructor the target bitrate, bus length(optional), transceiver timings (optional) and number of nodes (optional) and the software will calculate the required registers needed to make it work. Optionally There can be BRP, TSEG1, TSEG2, SJW and sample point parameters can be used

You make a very good point, fully agree. I note @kwagyeman said similar, too. Regarding exposing non-hardware-specific constructor options for bus length and transceiver timings: do you know if most manufacturers provide reasonable algorithms in order to calculate low-level configuration register values from these parameters? If so, then it seems reasonable to require drivers to expose these as common arguments. If not, then falling back to controller-specific register arguments may be the best (i.e. minimal cross-hardware common settings, maximum hardware-specific flexibility). I'll dig into this myself as well.

> There should also be a "Message" class. This class represents a CAN frame.  [...]
> I recommend splitting the CAN implementation into 2 pieces. the first piece is what the user interacts with. it is the classes. The other is the bus. The only exposed portion of the bus is the constructor. The user will not have access to any other portion of the bus.

Unless I'm missing something then this is essentially the design proposed in this PR. The part in the `machine.CAN.rst` file is the low level hardware-specific API, and the idea is to ship a high-level Pythonic API in micropython-lib that is what most folks will interact with (the sample code in the description of this PR is to give a taste of that API).

We don't specifically disallow using the lower level classes directly (it is Python not C++ after all), but we'll try to ship the high level classes at the same time, document them well, etc. The question mentioned above of whether or not to include a simple `recv()` in the low-level API at all is probably the outstanding question of how much we want to "nudge" versus "shove" people towards always using the high-level classes.

If you think I've missed another distinction compared to the design you've described, please let me know.

> There should be 2 distinct classes for the CAN. It should be CAN and CANFD. they should not be combined.

I'm not sure I follow the reasoning here, especially as CAN and CAN-FD nodes can exist on the same bus so the controller needs to be aware of both protocols for many applications. Can you please explain your reasoning some more, in particular the pitfalls of a combined design? Particularly if most users are going to be working with a higher level common API anyhow (if I understand your comments correctly, you're saying that the higher layer API should be combined.)

projectgus commented:
@IhorNehrutsa @obdevel @andrewleech @kentindell Do any of you have time to take a look at this proposed API? Additional feedback or questions would be very welcome! Cheers.

kentindell commented:
A few comments.

- Very pleased to see a priority queue. Nearly everyone gets this wrong and breaks their CAN implementation as a result. But there is also a need to feed a priority queue from one or more FIFO queues (ISO TP messages will need this, for example). It's better to define the way to do that than have everyone roll their own.
- On the timing properties of CAN, you shouldn't take a cue from what the hardware does, but rather specify exactly the parameters defined in ISO 11898. These are complicated, and people just want to get started, so define the common bit rates as profiles and let people choose one. This has the benefit of making it harder to screw up (people always get the sample point in the wrong place because they're thinking of UARTs and reckon 50% is good - which it very much isn't). I really wouldn't try to calculate these fields dynamically: it's a complex problem. Better to have simple quickstart settings and then anyone who wants to go off-route can do their own thing with ISO standard parameters.
- The above goes double for CAN FD!
- ISO 11898 defines some options that the hardware may choose to support, like disabling error re-transmission or ignoring CAN FD frames. I would include those as defined options because they're standard ones, and then ignore them where the hardware doesn't support them.
- I don't think you'll get away with Python only for the low level drivers for 3 TX buffer controllers like the bxCAN. Building a priority queue in software requires the ISR responding to the TX / Abort interrupt to juggle buffers very quickly. But I guess we'll have to see.
- It would be nice to support time-stamping in a standard way: most modern hardware does it, and it's needed for someone doing time-triggered CAN in software (I don't think TTCAN solves real-time problems, but that's another story).

andrewleech commented:
>  Do any of you have time to take a look at this proposed API? 

Unfortunately I've got very little actual knowledge about can usage, my interest is more in getting canopen working eventually to broaden the appeal of micropython at work :-D

projectgus commented:
Thanks @kentindell for weighing in here.

> Very pleased to see a priority queue. Nearly everyone gets this wrong and breaks their CAN implementation as a result. But there is also a need to feed a priority queue from one or more FIFO queues (ISO TP messages will need this, for example). It's better to define the way to do that than have everyone roll their own.

In the lower layer, that's the intention of the `send(..., fifo_equal=True)` parameter. By default, send shouldn't requeue a message of equal priority and the caller will be able to get FIFO transmission. To requeue a newer message with equal priority over the top of one already in the hardware buffer requires setting `fifo_equal=False`. However this doesn't account for the possibility that you might want FIFO transmission of messages with different priorities (is that ever a thing?)

On the higher layer, the "FIFO send" API is not yet defined. For ergonomics, it might need to be made more explicit with a separate queue or a different API. 

How does that sound?

> On the timing properties of CAN, you shouldn't take a cue from what the hardware does, but rather specify exactly the parameters defined in ISO 11898. [..]
> The above goes double for CAN FD!

Understood. See [the threaded discussion above](https://github.com/micropython/micropython/pull/13149#discussion_r1448089343) for more on this one.

> ISO 11898 defines some options that the hardware may choose to support, like disabling error re-transmission or ignoring CAN FD frames. I would include those as defined options because they're standard ones, and then ignore them where the hardware doesn't support them.

Thanks, will look at these.

>  I don't think you'll get away with Python only for the low level drivers for 3 TX buffer controllers like the bxCAN. Building a priority queue in software requires the ISR responding to the TX / Abort interrupt to juggle buffers very quickly. But I guess we'll have to see.

Agree, we'll see. I'm confident the higher level Python ISRs won't need to allocate, so on STM32 this means they can be a "hard" ISR. But if we end up needing to implement the software priority queue in C, that's fine (it can still be a generic piece of code shared by all drivers, which is a win.)

> It would be nice to support time-stamping in a standard way: most modern hardware does it, and it's needed for someone doing time-triggered CAN in software (I don't think TTCAN solves real-time problems, but that's another story).

Do you mean on the sending side, so you could schedule a message to send in a particular window for TTCAN-like functions? This feels like it's best to leave above the individually controller driver layer, even if that reduces timing tolerance a bit. (From what little I know about TTCAN it seems like the sort of thing where you probably need a harder real-time system than MicroPython to pull it off, anyway.) 

On the receive side, what's written in there now is to have all drivers timestamp incoming messages with `time.ticks_us()`, and the driver can do its best to accurately do that translation. Do you think that's precise enough? In Python applications it's pretty difficult to do anything with much finer timing.

kentindell commented:
I think microsecond accuracy timestamps are fine (and a natural human readable unit). I just hardwired the timestamps in the CANPico drivers to this. It's enough accuracy to spot crystal clock drift between devices (the reason for collecting the transmit timestamp too is that you can send a "follow up" message of that timestamp, and devices can sync their local clocks and then coordinate actions at the same time).

I don't think it's worth supporting FIFOs with different IDs, by the way: horrible side effects and don't think it's helpful. But multiple FIFOs into a single priority transmit queue is important (multiple ISO TP messages, and MQTT over CAN).

Someone mentioned using the HAL to do the drivers. I can tell you that's a terrible idea: I've never seen a HAL drive CAN properly. They routinely suck, almost no-one who writes CAN drivers has heard about priority inversion, and I've never seen anyone write bxCAN drivers properly (I have some proper bxCAN drivers you can have if you want: supports up to 32 frames in a priority queue in RAM).

My other advice is not to separate FD from CAN. No-one using MicroPython is going to worry about the buffer sizing issues, and you can in any case use what the CAN hardware provides (in the 2518FD hardware I just make the priority transmit queue have a depth of 20). For the receive FIFO it's just a FIFO and different frame sizes aren't a problem.

When CAN XL hardware turns up with its 2048 byte frames, then we can revisit this problem.

kdschlosser commented:
to give a basic idea of how to go about doing this where the front end API would remain a constant across all MCU's

[CAN.zip](https://github.com/micropython/micropython/files/13896544/CAN.zip)

@kentindell 
What you may believe people will and will not do is an assumption. Being wasteful with resources no matter what the platform or machine is just plain ol bad programming abilities. MicroPython goes on device that have < 400K of RAM. wasting memory is one of those things that can cause a lot of headaches. You saying that makes me think that you are young and you have never had to deal with computers that only had 640K of memory, You never had to experience the headache of getting drivers loaded in that space and still having enough left over to run an application. OYE! I can tell you it was not enjoyable that's for sure. Someone that says "a few bytes don't matter" has never had a situation where they were short a few bytes of memory to be able to run an application.

If everything in MicroPython was written using that thought process it would not be able to run on the MCU's it runs on.

Add a GUI interface on the MCU like LVGL along with CAN and see how quickly the memory gets gobbled up.  So yes every single byte does count.

projectgus commented:
> Being wasteful with resources no matter what the platform or machine is just plain ol bad programming abilities

@kdschlosser Please refrain from disparaging others' programming abilities, it's unacceptable and unproductive.

MicroPython priorities RAM usage heavily, and there are more than two possible ways to approach this. It should be entirely possible to have a single controller driver that pre-allocates buffers of different sizes depending on whether or not CAN FD is in use.

kdschlosser commented:
OK so please explain in what way did I disparage someone's programming abilities? by saying the comment has led to me believe they were young? making the statement about wasting resources is a true statement. I would hope that it is not being done deliberately and if it's not then it must be from lack of ability/experience right? 

 Maybe if i say someone's idea sucks is better?

projectgus commented:
I've edited this thread to remove some comments. MicroPython has a [code of conduct](https://github.com/micropython/micropython/blob/master/CODEOFCONDUCT.md), and we will enforce it if necessary. Making constructive comments about the specific technical merits (or not) of an idea is very welcome, but it's essential to do so respectfully.

projectgus commented:
I've made some more updates based on the feedback here, thanks everyone for your time looking over this doc.

Next will try implementing an initial driver. Will use the STM32 CAN controller as we have the existing pyboard implementation.

projectgus commented:
@kentindell Yes, for sure - I should have given more details. I'm hoping we can make the hardware controller specific parts a thin wrapper around the hardware queue, and the software queueing mechanism will be in the layer above (which will be Python). If that's too slow then we'll bump the software queueing part into C, but implemented in an abstract way so that the same implementation can work across different hardware.

BTW, you mentioned that you have a "proper" bxCAN driver implementation that skips the common pitfalls. If you're able to give me a peek, please send it through - I'd be interested to look.

kwagyeman commented:
@projectgus - In the interest of having something versus nothing... are you ready to say the API is finalized? I'd like to not be in limbo moving forwards as we've got users wanting a stable CAN API.

projectgus commented:
> are you ready to say the API is finalized? I'd like to not be in limbo moving forwards as we've got users wanting a stable CAN API.

I think this is very close to what the Python API for low-level CAN controllers will be. I'd hesitate to say it's final until it's been tested in an implementation.

From the perspective of writing a driver, the question of where software queueing sits (Python, C) may change if we can't meet decent performance with this part written in Python. Moving that may require small changes to the Python API documented here, and likely change the controller driver implementation substantially (i.e. which parts are C code included from `extmod/` versus implemented in higher level Python).

From the user's perspective, the low-level controller API defined here is fairly bare bones and won't be particularly ergonomic until there's a high-level can module to sit on top of it.

So, on balance, I'd recommend waiting until there's an initial driver implementation available with PRs into both `micropython` and `micropython-lib` before refactoring any existing CAN drivers - just in case.

Tomblarom commented:
As a typical "python user" with basic hw-knownledge, that has the interest to use CAN on micropython it was captivated by reading where this implementation started and how it formed over the years, only to be finally merged soon. I really take my hat off to your work and look forward to being a user of your fruits soon. I hope my words motivate you a little!

IhorNehrutsa commented:
I drew attention to two code examples with
`machine.CAN.NORMAL` 
```
can = machine.CAN(0, bitrate=1_000_000, mode=machine.CAN.NORMAL)
```
and `machine.CAN.Mode.NORMAL`
```
can.reset(mode=machine.CAN.Mode.NORMAL)
```
I know how to do the first example with the one-level dictionary like in the `machine.Pin(4, mode=machine.Pin.OUT)` 
```
static const mp_rom_map_elem_t machine_pin_locals_dict_table[] = {
static const mp_rom_map_elem_t machine_pin_locals_dict_table[] = {
    // instance methods
    { MP_ROM_QSTR(MP_QSTR_init), MP_ROM_PTR(&machine_pin_init_obj) },
    { MP_ROM_QSTR(MP_QSTR_value), MP_ROM_PTR(&machine_pin_value_obj) },
....
    // class constants
    { MP_ROM_QSTR(MP_QSTR_IN), MP_ROM_INT(GPIO_MODE_INPUT) },
    { MP_ROM_QSTR(MP_QSTR_OUT), MP_ROM_INT(GPIO_MODE_INPUT_OUTPUT) },
    { MP_ROM_QSTR(MP_QSTR_OPEN_DRAIN), MP_ROM_INT(GPIO_MODE_INPUT_OUTPUT_OD) },
...
    { MP_ROM_QSTR(MP_QSTR_PULL_UP), MP_ROM_INT(GPIO_PULL_UP) },
    { MP_ROM_QSTR(MP_QSTR_PULL_DOWN), MP_ROM_INT(GPIO_PULL_DOWN) },
...
    { MP_ROM_QSTR(MP_QSTR_DRIVE_0), MP_ROM_INT(GPIO_DRIVE_CAP_0) },
    { MP_ROM_QSTR(MP_QSTR_DRIVE_1), MP_ROM_INT(GPIO_DRIVE_CAP_1) },
    { MP_ROM_QSTR(MP_QSTR_DRIVE_2), MP_ROM_INT(GPIO_DRIVE_CAP_2) },
    { MP_ROM_QSTR(MP_QSTR_DRIVE_3), MP_ROM_INT(GPIO_DRIVE_CAP_3) },
};
```
Is there an example of a two-level dictionary? Like the `machine.CAN.Mode.NORMAL`

projectgus commented:
@IhorNehrutsa There is a way to do this, but I'd recommend not trying to implement this spec until I've got the stm32 implementation finished. Some things will change, possibly even including this two-level enum pattern.

IhorNehrutsa commented:
Considering merge of **stm32: Refactor pyb.CAN implementation, add tests.** #15989 what is the status of this PR?

projectgus committed:
docs: Update machine.CAN docs.

Changes made so far while developing the stm32 machine.CAN driver.

Still work-in-progress.

This work was funded through GitHub Sponsors.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

projectgus commented:
> Considering merge of **stm32: Refactor pyb.CAN implementation, add tests.** #15989 what is the status of this PR?

This docs PR was created as a discussion point for an API, but it probably won't be merged until there is a `machine.CAN` implementation for stm32 - there's no point generating docs for an API we don't support yet. I have made a few adjustments while working on `machine.CAN` for stm32, have just pushed those here.

Still recommend not implementing any driver around this yet, it will probably change more (and the stm32 implementation will add generic code in the `extmod/` directory that the port drivers will be expected to use.)

Thanks for being patient. A draft PR for stm32 is inching closer, as available time permits.

jrborbars commented:
Hey guys, any news here 9 months after the last post? I'm really interested in this API and can try to help, if it is still needed.

projectgus commented:
Sorry everyone, this work has taken a lot longer than anyone anticipated. Hard to believe I've been working on it on-and-off for two years, but I have. Was consistently bogged down due to a couple of things:

1. The draft API documented here is too complex, especially the CAN TX logic. In trying to get "ideal" CAN node behaviour I made it so co-ordinating state between the Python code and the C driver was very hard under some edge cases. In particular there's an inherent race condition if the `send()` function "cancels" sending of a particular message, replacing it with a higher priority message, and then the TX IRQs run - which message does the IRQ refer to?
2. The STM32 CAN controller peripherals don't seem to behave as documented when there are multiple outgoing messages with the same CAN ID. There is a documented ordering (lowest numbered buffer transmits first), but the controller sometimes picks a random order. [I'm not the only person to notice this](https://forum.opencyphal.org/t/uavcan-can-tx-buffer-management-in-can-fd-controllers/1215/2), but it took me *way* too long to realise it wasn't a bug I'd written (refer to point 1 about the design being too complex.)

So I went through many cycles of "refactor CAN TX implementation, tests fail, try to debug, get busy with other work, repeat". I think I re-wrote the implementation three or four times before deciding the design was the problem (and then eventually realising that the hardware was also a problem!) Thanks to Damien who suggested taking the "de-queue a lower priority message" behaviour out of the `CAN.send()` function entirely.

Anyway, a new PR is open - https://github.com/micropython/micropython/pull/18572 - which has an stm32 `machine.CAN` driver and new API docs. Thanks everyone for being so patient. I'm going to close this PR in favour of that one.
